packages/grafana-ui/src/components/Sparkline/Sparkline.tsx

```diff
@@ -0,0 +1,171 @@
+      ...sparkline.y.config,
+      ...this.props.config,
+    };
+
+    return {
+      refId: 'sparkline',
+      fields: [
+        sparkline.x ?? IndexVector.newField(length),
+        {
+          ...sparkline.y,
+          config: yFieldConfig,
+        },
+      ],
+      length,
+    };
+  }
+
+  prepareConfig(data: DataFrame, props: Props) {
+    const { theme } = this.props;
+    const builder = new UPlotConfigBuilder();
+
+    builder.setCursor({
+      show: true,
+      x: false, // no crosshairs
+      y: false,
+    });
+
+    // X is the first field in the alligned frame
+    const xField = data.fields[0];
+    builder.addScale({
+      scaleKey: 'x',
+      isTime: false, //xField.type === FieldType.time,
+      range: () => {
+        const { sparkline } = this.props;
+        if (sparkline.x) {
+          if (sparkline.timeRange && sparkline.x.type === FieldType.time) {
+            return [sparkline.timeRange.from.valueOf(), sparkline.timeRange.to.valueOf()];
+          }
+          const vals = sparkline.x.values;
+          return [vals.get(0), vals.get(vals.length - 1)];
+        }
+        return [0, sparkline.y.values.length - 1];
+      },
+    });
+
+    builder.addAxis({
+      scaleKey: 'x',
+      theme,
+      placement: AxisPlacement.Hidden,
+    });
+
+    for (let i = 0; i < data.fields.length; i++) {
+      const field = data.fields[i];
+      const config = field.config as FieldConfig<GraphFieldConfig>;
+      const customConfig: GraphFieldConfig = {
+        ...defaultConfig,
+        ...config.custom,
+      };
+
+      if (field === xField || field.type !== FieldType.number) {
+        continue;
+      }
+
+      const scaleKey = config.unit || '__fixed';
+      builder.addScale({ scaleKey, min: field.config.min, max: field.config.max });
+      builder.addAxis({
+        scaleKey,
+        theme,
+        placement: AxisPlacement.Hidden,
+      });
+
+      const colorMode = getFieldColorModeForField(field);
+      const seriesColor = colorMode.getCalculator(field, theme)(0, 0);
+      const pointsMode = customConfig.drawStyle === DrawStyle.Points ? PointVisibility.Always : customConfig.showPoints;
+      builder.addSeries({
+        scaleKey,
+        drawStyle: customConfig.drawStyle!,
+        lineColor: customConfig.lineColor ?? seriesColor,
+        lineWidth: customConfig.lineWidth,
+        lineInterpolation: customConfig.lineInterpolation,
+        showPoints: pointsMode,
+        pointSize: customConfig.pointSize,
+        pointColor: customConfig.pointColor ?? seriesColor,
+        fillOpacity: customConfig.fillOpacity,
+        fillColor: customConfig.fillColor ?? seriesColor,
+      });
+    }
+
+    return builder;
+  }
+
+  render() {
+    const { data, configBuilder } = this.state;
+    const { width, height, sparkline } = this.props;
+
+    return (
+      <UPlotChart
+        data={{
+          frame: data,
+          isGap: () => true, // any null is a gap
+        }}
```

Comment:
this is not ideal, creating a new data object here will make the Plot think the data always has updated. But think it's fine also as this will only be in-efficient when resizing the sparkline (only other reason for the sparkline to re-render in cases when data not changed is resize as far as I can see).  So this is not a problem really.